openedx_certificates/models.py

```diff
@@ -1,5 +1,345 @@
+
+        Args:
+            user_id: The ID of the user to process the certificate for.
+            celery_task_id (optional): The ID of the Celery task that is running this function.
+        """
+        user = get_user_model().objects.get(id=user_id)
+        # Use the name from the profile if it is not empty. Otherwise, use the first and last name.
+        # We check if the profile exists because it is absent in unit tests.
+        user_full_name = getattr(getattr(user, 'profile', None), 'name', f"{user.first_name} {user.last_name}")
+        custom_options = {**self.certificate_type.custom_options, **self.custom_options}
+
+        try:
+            certificate = ExternalCertificate.objects.get(
+                user_id=user_id,
+                course_id=self.course_id,
+                certificate_type=self.certificate_type.name,
+            )
+            certificate.user_full_name = user_full_name
+            certificate.status = ExternalCertificate.Status.GENERATING
+            certificate.generation_task_id = celery_task_id
+            certificate.save()
+        except ExternalCertificate.DoesNotExist:
+            certificate = ExternalCertificate.objects.create(
+                user_id=user_id,
+                user_full_name=user_full_name,
+                course_id=self.course_id,
+                certificate_type=self.certificate_type.name,
+                status=ExternalCertificate.Status.GENERATING,
+                generation_task_id=celery_task_id,
+            )
+
+        try:
+            generation_module_name, generation_func_name = self.certificate_type.generation_func.rsplit('.', 1)
+            generation_module = import_module(generation_module_name)
+            generation_func = getattr(generation_module, generation_func_name)
+
+            # Run the functions. We do not validate them here, as they are validated in the model's clean() method.
+            certificate.download_url = generation_func(self.course_id, user, certificate.uuid, custom_options)
+            certificate.status = ExternalCertificate.Status.AVAILABLE
+            certificate.save()
+        except Exception as exc:  # noqa: BLE001
+            certificate.status = ExternalCertificate.Status.ERROR
+            certificate.save()
+            msg = f'Failed to generate the {certificate.uuid=} for {user_id=} with {self.id=}.'
+            raise CertificateGenerationError(msg) from exc
+
+
+class ExternalCertificate(TimeStampedModel):
+    """
+    Model to represent each individual certificate awarded to a user for a course.
+
+    This model contains information about the related course, the user who earned the certificate,
+    the download URL for the certificate PDF, and the associated certificate generation task.
+
+    .. note:: The ID field is not a conventional auto-incrementing integer, but a value
+       that allows for old certificates with custom IDs.
```

Comment:
```suggestion
    .. note:: Certificates are identified by UUIDs rather than integer keys. Not only
       does this prevent enumeration attacks / privacy leaks, but it is backwards-
       compatible with old certificate formats, so we can import and represent them
       here without changing their IDs.
```
I was slightly confused that there is no `ID` field here, because it seems pretty obvious that a field called `uuid` is not going to be an integer, so I thought maybe you were talking about some other field. Or are you talking about `legacy_id`?